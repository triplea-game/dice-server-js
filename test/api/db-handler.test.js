jest.enableAutomock();
jest.unmock('../../src/api/db-handler');

const pg = require('pg-promise');

const mockedPg = jest.fn();
pg.mockReturnValue(mockedPg);
const DbHandler = require('../../src/api/db-handler');

const dbObject = {
  none: jest.fn(),
  result: jest.fn(),
  oneOrNone: jest.fn(),
};

describe('The DB-Handler', () => {
  beforeAll(() => {
    mockedPg.mockReturnValue(dbObject);
    dbObject.none.mockReturnValue(Promise.resolve(true));
  });

  it('should create a database when initialized', () => {
    const handler = new DbHandler({
      username: 'user',
      password: 'secret',
      host: '[::1]',
      port: '1337',
      database: 'someDb',
    });

    handler.setupDb();

    expect(mockedPg).toMatchSnapshot();
    expect(dbObject.none).toMatchSnapshot();
  });

  it('should log an error if something goes wrong', async () => {
    global.console.error = jest.fn();
    dbObject.none = jest.fn();
    const testError = new Error('ERROR MESSAGE');
    dbObject.none.mockReturnValue(Promise.reject(testError));
    await new DbHandler({}).setupDb();

    expect(global.console.error).toHaveBeenCalledTimes(2);
    expect(global.console.error).toHaveBeenLastCalledWith(testError);
    global.console.error.mockRestore();
  });

  it('should execute the correct call to the database on addUser', () => {
    const handler = new DbHandler({});
    handler.addUser('Test Email');
    expect(dbObject.none).toMatchSnapshot();
  });

  it('should execute the correct call to the database on removeUser', () => {
    const handler = new DbHandler({});
    handler.removeUser('Test Email');
    expect(dbObject.result).toMatchSnapshot();
  });

  it('should execute the correct call to the database on checkMail', () => {
    const handler = new DbHandler({});
    handler.checkMail('Test Email');
    expect(dbObject.oneOrNone).toMatchSnapshot();
  });
});
