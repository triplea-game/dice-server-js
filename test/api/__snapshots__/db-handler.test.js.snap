// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DB-Handler should create a database when initialized 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "postgres://user:secret@[::1]:1337/someDb",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "none": [MockFunction] {
          "calls": Array [
            Array [
              "CREATE TABLE IF NOT EXISTS users (email varchar(65) NOT NULL PRIMARY KEY);",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": Promise {},
            },
          ],
        },
        "oneOrNone": [MockFunction],
        "result": [MockFunction],
      },
    },
  ],
}
`;

exports[`The DB-Handler should create a database when initialized 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "CREATE TABLE IF NOT EXISTS users (email varchar(65) NOT NULL PRIMARY KEY);",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`The DB-Handler should execute the correct call to the database on addUser 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "INSERT INTO users (email) VALUES ($1)",
      "Test Email",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`The DB-Handler should execute the correct call to the database on checkMail 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "SELECT email FROM users WHERE email=$1",
      "Test Email",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`The DB-Handler should execute the correct call to the database on removeUser 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "DELETE FROM users WHERE email=$1",
      "Test Email",
      [Function],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
